pa4/checker-pa4.ml

```diff
@@ -428,31 +450,39 @@ let rec exp_typecheck(o: obj_env)  (exp: expression) : static_type = begin
                 printf "ERROR: %s: Type-check: undeclared variable %s\n"  id_location id_name;
                 exit 1
         end
-
-	(*| Let((vloc, vname), (typeloc, typename), None, let_body)-> 
-		(*add vname to O -- add it to current scope *)
-		Hashtbl.add o vname (Class typename) ; (* TODO: SELF_TYPE? *)
-		(*We check the let body with bound variable added to obj env*)
-		let body_type = exp_typecheck o let_body in
-		(*remove vname from O -- after the current scope *)
-		Hashtbl.remove o vname;
-		body_type;
-	| Let((vloc, vname), (typeloc, typename), Some(init_exp), let_body)-> 
-		(* TODO: SELF_TYPE? *)
-		(* init exp - subtype check*)
-		let init_type = exp_typecheck o init_exp in
-		if not (is_subtype init_type (Class typename)) then begin
-			printf "ERROR: %s: Type-Check: initializer for %s was %s, did not match declared %s\n" 
-				exp.line_number vname (type_to_str init_type) typename ;
-			exit 1
-		end;
-		(*add vname to O -- add it to current scope *)
-		Hashtbl.add o vname (Class typename) ; (* TODO: SELF_TYPE? *)	
-		(*We chek the let body with bound variable added to obj env*)
-		let body_type = exp_typecheck o let_body in
-		(*remove vname from O -- after the current scope *)
-		Hashtbl.remove o vname;
-		body_type;*)
+	| If(predicate, then_exp, else_exp) ->
+    	let predicate_type = exp_typecheck o_e m_e predicate in
+    	if predicate_type <> (Class "Bool") then begin
+    		printf "ERROR: %s: If statement's predicate expects type Bool, not type %s\n" exp.line_number (type_to_str predicate_type);
+    		exit 1
+    	end
+    	else
+    		let then_exp_type = exp_typecheck o_e m_e then_exp in 
+    		let else_exp_type = exp_typecheck o_e m_e else_exp in
+    			lowest_upper_bound then_exp_type else_exp_type
+    | Block(exps) ->
+    	let list_type = ref (Class "Object") in 
+    	let check_type = List.iter(fun e -> 
+        		list_type := exp_typecheck o_e m_e e
+    	) exps in 
+    		!list_type
+    | Let(bindings, exp) ->
+    	let new_o = o_e in  
+		List.iter(fun binding -> 
+			match binding with 
+			| BindingNoInit((id_location, id_name), (type_location, init_type)) -> 
+				Hashtbl.add new_o id_name (Class init_type)
+			| BindingInit((id_location, id_name), (type_location, init_type), init_exp) ->
+				let init_exp_type = exp_typecheck o_e m_e init_exp in 
+					if is_subtype init_exp_type (Class init_type) then 
```

Comment:
There is no reason for this line to be indented, it's actually confusing since it looks like we've entered something like a loop here.